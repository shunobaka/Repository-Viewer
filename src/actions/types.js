export const USER_LOADED = 'USER_LOADED';
export const USERS_SEARCH_RESULT = 'USERS_SEARCH_RESULT';
export const REPOSITORIES_LOADED = 'REPOSITORIES_LOADED';
export const REPOSITORIES_FILTERED = 'REPOSITORIES_FILTERED';
export const ADD_ALERT = 'ADD_ALERT';
