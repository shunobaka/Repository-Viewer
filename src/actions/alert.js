/**
 * @fileoverview Defines functions that return action objects used by redux to
 *    update alert state.
 */
import { ADD_ALERT, REMOVE_ALERT, CLEAR_ALERTS } from './types';

/**
 * Returns a redux action object based on the provided alert object.
 * @param {object} alert An object representing an alert with fields message and id
 * @return {object} Redux action object
 * @export
 */
export const addAlert = (alert) => {
  return {
    type: ADD_ALERT,
    payload: alert,
  };
};

/**
 * Returns a redux action object used to remove the alert with the provided id.
 * @param {string} id The uuid of the alert that is to be removed
 * @return {object} Redux action object
 * @export
 */
export const removeAlert = (id) => {
  return {
    type: REMOVE_ALERT,
    payload: id,
  };
};

/**
 * Returns a redux action object used to remove all alerts.
 * @return {object} Redux action object
 * @export
 */
export const clearAlerts = () => {
  return {
    type: CLEAR_ALERTS,
  };
};
