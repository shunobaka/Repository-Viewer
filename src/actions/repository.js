/**
 * @fileoverview Defines functions that return action objects used by redux to
 *    update repository state.
 */
import { REPOSITORIES_LOADED, REPOSITORIES_FILTERED } from './types';

/**
 * Returns a redux action object used to save repositories in state based on the
 *    provided repositories.
 * @param {array} repos Repository objects to be saved in redux state
 * @return {object} Redux action object
 * @export
 */
export const repositoriesLoaded = (repos) => {
  return {
    type: REPOSITORIES_LOADED,
    payload: repos,
  };
};

/**
 * Returns a redux action object used to filter repositories in state based on the
 *    provided query.
 * @param {string} query The name query by which repositories are filtered
 * @return {object} Redux action object
 * @export
 */
export const repositoriesFiltered = (query) => {
  return {
    type: REPOSITORIES_FILTERED,
    payload: query,
  };
};
