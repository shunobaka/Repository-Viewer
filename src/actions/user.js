/**
 * @fileoverview Defines functions that return action objects used by redux to
 *    update user state.
 */
import { USERS_SEARCH_RESULT, USER_LOADED } from './types';

/**
 * Returns a redux action object used to save users and username query in state
 *    based on the provided users and query.
 * @param {array} users User objects to be saved in the redux state
 * @param {*} query The username query by which the users were retrieved
 * @returns Redux action object
 * @export
 */
export const usersLoaded = (users, query) => {
  return {
    type: USERS_SEARCH_RESULT,
    payload: {
      users,
      query,
    },
  };
};

/**
 * Returns a redux action object used to save user in state based on the provided
 *    user.
 * @param {object} user User object to be saved in redux state
 * @returns Redux action object
 * @export
 */
export const userLoaded = (user) => {
  return {
    type: USER_LOADED,
    payload: user,
  };
};
