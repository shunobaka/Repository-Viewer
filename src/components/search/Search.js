import React, { Fragment } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import userActionCreator from '../../actionCreators/user';
import Users from './Users';

const Search = ({ match }) => {
  // TODO: Implement users fetching
  const users = [];

  return (
    <Fragment>
      <h2 className="text-center green-text">
        Search results for {match.params.query}:
      </h2>
      {users && users.count > 0 ? (
        <Users users />
      ) : (
        <h2 className="text-center green-text">
          No users found with this username. Please try again.
        </h2>
      )}
    </Fragment>
  );
};

const mapStateToProps = (state) => ({
  users: state.users,
});

const mapDispatchToProps = (dispatch) => {
  return { actions: bindActionCreators(userActionCreator, dispatch) };
};

export default connect(mapStateToProps, mapDispatchToProps)(Search);
