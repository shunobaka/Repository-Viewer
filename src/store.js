/**
 * @fileoverview Creates and exports the redux store.
 */
import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';

import combinedReducer from './reducers';

const middleware = [thunk];

/**
 * The redux store that is created and exported.
 * @export
 */
const store = createStore(
  combinedReducer,
  composeWithDevTools(applyMiddleware(...middleware))
);

export default store;
