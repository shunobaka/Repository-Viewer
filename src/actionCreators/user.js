import githubApi from '../utils/githubApi';
import { usersLoaded, userLoaded } from '../actions/user';
import { setAlert } from '../actionCreators/alert';

const getUsers = (nameQuery) => async (dispatch) => {
  try {
    const res = await githubApi.get(`/search/users?q=${nameQuery}`);

    const payload = res.data.items.map((item) => ({
      username: item.login,
      avatar: item.avatar_url,
      id: item.id,
      num_repos: item.public_repos,
    }));

    dispatch(usersLoaded(payload, nameQuery));
  } catch (err) {
    dispatch(
      setAlert(
        'There was a problem searching for users, please try again.',
        false
      )
    );
  }
};

const loadUser = (username) => async (dispatch) => {
  try {
    const res = await githubApi.get(`/users/${username}`);

    const user = {
      id: res.data.id,
      username: res.data.login,
      avatar: res.data.avatar_url,
      num_repos: res.data.public_repos,
    };

    dispatch(userLoaded(user));
  } catch (err) {
    dispatch(setAlert('Could not retrieve user, please try again.', false));
  }
};

export default {
  getUsers,
  loadUser,
};
