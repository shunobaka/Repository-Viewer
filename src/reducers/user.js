import { USER_LOADED, USERS_SEARCH_RESULT } from '../actions/types';

const initialState = {
  user: null,
  users: [],
  query: null,
  loading: true,
};

export default function (state = initialState, action) {
  const { type, payload } = action;

  switch (type) {
    case USER_LOADED:
      return {
        ...state,
        user: payload,
      };
    case USERS_SEARCH_RESULT:
      return {
        user: null,
        users: payload.users,
        query: payload.query,
        loading: false,
      };
    default:
      return state;
  }
}
