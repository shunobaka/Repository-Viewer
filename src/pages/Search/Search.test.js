import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer, { act } from 'react-test-renderer';
import { Provider } from 'react-redux';
import Search from './Search';
import store from '../../store';
import { usersLoaded } from '../../actions/user';
import userActionCreators from '../../actionCreators/user';

/** Mock functions to prevent creating http requests */
jest.mock('../../actionCreators/user.js');

userActionCreators.getUsers.mockImplementation(() => {
  return { type: 'mock_action' };
});

describe('search snapshots', () => {
  const query = 'shunobaka';

  const mockUsers = [
    {
      username: 'test1',
      id: 0,
      avatar: 'test1',
    },
    {
      username: query,
      id: 1,
      avatar: query,
      num_repos: 0,
    },
  ];

  const match = {
    params: {
      query,
    },
  };

  it('multiple users renders correctly', () => {
    act(() => {
      store.dispatch(usersLoaded(mockUsers, query));
    });

    const tree = renderer
      .create(
        <BrowserRouter>
          <Provider store={store}>
            <Search match={match} />
          </Provider>
        </BrowserRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('single user renders correctly', () => {
    act(() => {
      store.dispatch(usersLoaded([mockUsers[0]], query));
    });

    const tree = renderer
      .create(
        <BrowserRouter>
          <Provider store={store}>
            <Search match={match} />
          </Provider>
        </BrowserRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('no users renders correctly', () => {
    act(() => {
      store.dispatch(usersLoaded([], query));
    });

    const tree = renderer
      .create(
        <BrowserRouter>
          <Provider store={store}>
            <Search match={match} />
          </Provider>
        </BrowserRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
