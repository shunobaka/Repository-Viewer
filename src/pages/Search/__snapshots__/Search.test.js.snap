// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search snapshots multiple users renders correctly 1`] = `
Array [
  <h2
    className="text-center green-text"
  >
    Search results for 
    shunobaka
    :
  </h2>,
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-md-3"
      >
        <div
          className="jumbotron"
        >
          <a
            href="/user/test1"
            onClick={[Function]}
          >
            <h4
              className="text-center"
            >
              test1
            </h4>
            <img
              className=""
              src="test1"
              width="100%"
            />
          </a>
        </div>
      </div>
      <div
        className="col-md-3"
      >
        <div
          className="jumbotron"
        >
          <a
            href="/user/shunobaka"
            onClick={[Function]}
          >
            <h4
              className="text-center"
            >
              shunobaka
            </h4>
            <img
              className=""
              src="shunobaka"
              width="100%"
            />
          </a>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`search snapshots no users renders correctly 1`] = `
Array [
  <h2
    className="text-center green-text"
  >
    Search results for 
    shunobaka
    :
  </h2>,
  <div
    className="text-center container"
  >
    <br />
    <div
      className="alert alert-warning"
    >
      No users found matching 
      shunobaka
      !
    </div>
  </div>,
]
`;

exports[`search snapshots single user renders correctly 1`] = `
Array [
  <h2
    className="text-center green-text"
  >
    Search results for 
    shunobaka
    :
  </h2>,
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-md-3"
      >
        <div
          className="jumbotron"
        >
          <a
            href="/user/test1"
            onClick={[Function]}
          >
            <h4
              className="text-center"
            >
              test1
            </h4>
            <img
              className=""
              src="test1"
              width="100%"
            />
          </a>
        </div>
      </div>
    </div>
  </div>,
]
`;
